protocol/l1-upgrades.md

```diff
@@ -0,0 +1,305 @@
+
+## A single `SuperchainProxyAdmin`
+
+The existing L1 `ProxyAdmin` (of which there is currently one per OP Chain), will be replaced with a
+single L1 `SuperchainProxyAdmin`. The `SuperchainProxyAdmin`, in contrast with the L1 `ProxyAdmin`,
+will have minimal storage, and will not internally track the type of Proxy which each contract is.
+
+This contract will be owned by the [Upgrade
+Controller](https://github.com/ethereum-optimism/specs/blob/main/specs/protocol/stage-1.md#L30), and
+will be a simple pass through contract with a minimal interface.
+
+```
+/// @notice forwards calldata to the specified address.
+///   Reverts unless called by owner.
+function forward(address, bytes) external;
+
+/// @notice forwards calldata to the specified addresses.
+///   Reverts unless called by owner.
+function forward(address[], bytes[]) external;
+```
+
+This architecture makes it possible to manage upgrade authorization for the whole superchain through
+a single storage value in a single contract, rather than having a separate contract per OP Chain.
+
+It also avoids the need for an `sload` to ready the `proxyType` mapping for each contract.
+
+## Identifying OP Chains by the SystemConfig
+
+Rather than tracking the contracts within the `ProxyAdmin`, we will use the `SystemConfig` as the
+unique identifier for each OPChain, and get the contract addresses from it. The custom logic
+required for the two non-standard proxy types will be handled in the OPCM contract, and does not
+require tracking the proxy type, as it is known already for each contract, ie. the only non-standard
+Proxies are the `L1StandardBridge` (`L1ChugsplashProxy`) and `L1CrossDomainMessenger`
+(`ResolvedDelegateProxy`).
+
+## The upgrade process
+
+As part of the release process, the associated implementation contracts and a new OPCM, will be deployed.
+
+The Upgrade Controller MUST be a Safe in order to atomically:
+
+1. transfer ownership of the `SuperchainProxyAdmin` to the new OPCM.
+2. call `OPCM.upgrade()` which will return ownership to the Safe immediately upon completing the
+   upgrade.
+
+Thus the high level logic for upgrading a contract should be roughly as follows:
+
+```js
+SuperchainProxyAdmin admin;
+ISystemConfigs[] public systemConfigs;
+
+Addresses immutable implementations;
+
+constructor(Addresses _implementations, ISystemConfigs[] _systemConfigs) {
+  implementations = _implementations;
+  for(uint i=0; i < _systemConfigs.length; i++){
+    systemConfigs.push(_systemConfigs[i])
+  }
+}
+
+function upgrade() public {
+  for(uint i=0; i< systemConfigs.length; i++) {
+    // Read the `Addresses` struct from each `SystemConfig` in the `systemConfigs` mapping.
+    // For each entry in the `Addresses` struct:
+      // 1. Call the `SuperchainProxyAdmin` to update the implementation to the `Populator` contract and call `Populator.upgrade()`.
+      // 2. Update the implementation to the final implementation contract.
```

Comment:
We talked about this a bit, another option to consider is the double-delegatecall approach: do a 1 time change of all proxy implementations to a beacon contract, which looks up the implementation. Then this step moves outside the loop

Reply:
If we're indifferent to gas costs them I'm down to do this.

Reply:
agreed we won't do the double proxy thing.